ang/civicase/ActivityActions--tags.js

```diff
@@ -0,0 +1,216 @@
+     * @param {String} saveButtonLabel
+     * @param {String} operation
+     * @param {Array} activities
+     */
+    function openTagsModal (model, title, saveButtonLabel, operation, activities) {
+      dialogService.open('TagsActivityAction', '~/civicase/ActivityActions--tags.html', model, {
+        autoOpen: false,
+        height: 'auto',
+        width: '40%',
+        title: title,
+        buttons: [{
+          text: saveButtonLabel,
+          icons: operation === 'add' ? {primary: 'fa-check'} : false,
+          click: function () {
+            addRemoveTagsConfirmationHandler.call(this, operation, activities, model);
+          }
+        }]
+      });
+    }
+
+    /**
+     * Add/Remove tags confirmation handler
+     *
+     * @param {String} operation
+     * @param {Array} activities
+     * @param {Object} model
+     * @return {Promise}
+     */
+    function addRemoveTagsConfirmationHandler (operation, activities, model) {
+      var tagIds = model.selectedGenericTags;
+
+      _.each(model.tagSets, function (tag) {
+        if (tag.selectedTags) {
+          tagIds = tagIds.concat(JSON.parse('[' + tag.selectedTags + ']'));
+        }
+      });
+
+      var apiCalls = prepareApiCalls(operation, activities, tagIds);
+
+      crmApi(apiCalls)
+        .then(function () {
+          $rootScope.$emit('civicase::activity::updated');
+        });
+
+      $(this).dialog('close');
+    }
+
+    /**
+     * Prepare the API calls for the Add/Remove operation
+     *
+     * @param {String} operation
+     * @param {Array} activities
+     * @param {Array} tagIds
+     * @return {Array}
+     */
+    function prepareApiCalls (operation, activities, tagIds) {
+      var apiCalls = [];
+      var action = operation === 'add' ? 'create' : 'delete';
+
+      _.each(activities, function (activity) {
+        apiCalls.push(['EntityTag', action, {
+          entity_table: 'civicrm_activity',
+          entity_id: activity.id,
+          tag_id: tagIds
+        }]);
+      });
+
+      return apiCalls;
+    }
+    /**
+     * Get the tags for Activities from API end point
+     *
+     * @return {Promise}
+     */
+    function getTags () {
+      return crmApi([['Tag', 'get', {
+        'sequential': 1,
+        'used_for': { 'LIKE': '%civicrm_activity%' }
+      }]]).then(function (data) {
+        return data[0].values;
+      });
+    }
+
+    /**
+     * Recursive function to prepare the generic tags
+     *
+     * @param {Array} tags
+     * @param {String} parentID
+     * @param {int} level
+     * @return {Array}
+     */
+    function prepareGenericTags (tags, parentID, level) {
+      var returnArray = [];
+
+      level = typeof level !== 'undefined' ? level : 0;
+      parentID = typeof parent !== 'undefined' ? parentID : undefined;
+
+      var filteredTags = _.filter(tags, function (child) {
+        return child.parent_id === parentID && child.is_tagset === '0';
+      });
+
+      if (!_.isEmpty(filteredTags)) {
```

Comment:
Can we do an early return and save an extra level of nesting?
Something like
```js
if (_.isEmpty(filteredTags)) {
 return;
}

_.each(filteredTags, function (tag) {
  returnArray.push(tag);
  tag.indentationLevel = level;
  returnArray = returnArray.concat(prepareGenericTags(tags, tag.id, level + 1));
});

return returnArray;
```

This will give more readability to the code.